src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/Formatting/FormattingLanguageServerClient.cs

```diff
@@ -104,60 +108,83 @@ private RazorDocumentRangeFormattingResponse Format(RazorDocumentRangeFormatting
+                var bufferManager = Activator.CreateInstance(bufferManagerType, new object[] { contentTypeService, textBufferFactoryService, textBufferListeners });
+                var joinableTaskFactoryThreadSwitcher = typeof(IdAttribute).Assembly.GetType("Microsoft.WebTools.Shared.Threading.JoinableTaskFactoryThreadSwitcher", throwOnError: true);
+                var joinableTaskFactoryThreadSwitcherCtor = joinableTaskFactoryThreadSwitcher.GetConstructor(new[] { typeof(JoinableTaskFactory) });
+                var threadSwitcher = (IThreadSwitcher)joinableTaskFactoryThreadSwitcherCtor.Invoke(new object[] { new JoinableTaskContext().Factory });
+                var applyFormatEditsHandler = Activator.CreateInstance(editHandlerType, new object[] { bufferManager, threadSwitcher, textBufferFactoryService });
+
+                // Make sure the buffer manager knows about the source document
+                var documentUri = DocumentUri.From($"file:///{@params.HostDocumentFilePath}");
+                var contentTypeName = HtmlContentTypeDefinition.HtmlContentType;
+                var initialContent = generatedHtml;
+                var snapshotVersionFromLSP = 0;
+                Assert.IsAssignableFrom<ITextSnapshot>(bufferManager.GetType().GetMethod("CreateBuffer").Invoke(bufferManager, new object[] { documentUri, contentTypeName, initialContent, snapshotVersionFromLSP }));
+
+                var requestType = editHandlerAssembly.GetType("Microsoft.WebTools.Languages.LanguageServer.Server.ContainedLanguage.ApplyFormatEditsParamForOmniSharp", throwOnError: true);
+                var serializedValue = $@"{{
+    ""Options"": {{
+        ""UseSpaces"": {(@params.Options.InsertSpaces ? "true" : "false")},
+        ""TabSize"": {@params.Options.TabSize},
+        ""IndentSize"": {@params.Options.TabSize}
+    }},
+    ""SpanToFormat"": {{
+        ""start"": {@params.ProjectedRange.AsTextSpan(generatedHtmlSource).Start},
+        ""length"": {@params.ProjectedRange.AsTextSpan(generatedHtmlSource).Length},
+    }},
+    ""Uri"": ""file:///{@params.HostDocumentFilePath}"",
+    ""GeneratedChanges"": [
+    ]
+}}
+";
+                var request = JsonConvert.DeserializeObject(serializedValue, requestType);
+                var resultTask = (Task)applyFormatEditsHandler.GetType().GetRuntimeMethod("Handle", new Type[] { requestType, typeof(CancellationToken) }).Invoke(applyFormatEditsHandler, new object[] { request, CancellationToken.None });
+                var result = resultTask.GetType().GetProperty(nameof(Task<int>.Result)).GetValue(resultTask);
+                var rawTextChanges = result.GetType().GetProperty("TextChanges").GetValue(result);
+                var serializedTextChanges = JsonConvert.SerializeObject(rawTextChanges, Newtonsoft.Json.Formatting.Indented);
+                var textChanges = JsonConvert.DeserializeObject<HtmlFormatterTextEdit[]>(serializedTextChanges);
```

Comment:
I'm surprised there's not a better edit type we can deserialize into

Reply:
There may be... can check.

Reply:
I couldn't find one.